prow/cluster_lib.sh

```diff
@@ -20,23 +20,147 @@
+             cluster: ${CLUSTER_NAME}
+             user: ${CLUSTER_NAME}
+           name: ${CLUSTER_NAME}
+      current-context: ${CLUSTER_NAME}
+      kind: Config
+      preferences: {}
+      users:
+         - name: ${CLUSTER_NAME}
+           user:
+             token: ${TOKEN}
+EOF
+
+}
+
+function setup_clusterreg () {
+    # setup cluster-registries dir setup by mason
+    CLUSTERREG_DIR=${CLUSTERREG_DIR:-$(mktemp -d /tmp/clusterregXXX)}
+
+    SERVICE_ACCOUNT=istio-multi-test
+
+    # mason dumps all the kubeconfigs into the same file but we need to use per cluster
+    # files for the clusterregsitry config.  Create the separate files.
+    #  -- if PILOT_CLUSTER not set, assume pilot install to be in the first cluster
+    PILOT_CLUSTER=${PILOT_CLUSTER:-$(kubectl config current-context)}
+    unset IFS
+    k_contexts=$(kubectl config get-contexts -o name)
+    for context in ${k_contexts}; do
+        if [[ "$PILOT_CLUSTER" != "$context" ]]; then
+            kubectl config use-context ${context}
+             
+            kubectl create ns ${SA_NAMESPACE}
+            kubectl create sa ${SERVICE_ACCOUNT} -n ${SA_NAMESPACE}
+            kubectl create clusterrolebinding istio-multi-test --clusterrole=cluster-admin --serviceaccount=${SA_NAMESPACE}:${SERVICE_ACCOUNT}
+            CLUSTER_NAME=$(kubectl config view --minify=true -o "jsonpath={.clusters[].name}")
+            if [[ "${CLUSTER_NAME}" =~ .*"_".* ]]; then
+                # if clustername has '_' set value to stuff after the last '_' due to k8s secret data name limitation
+                CLUSTER_NAME=${CLUSTER_NAME##*_}
+            fi
+            KUBECFG_FILE=${CLUSTERREG_DIR}/${CLUSTER_NAME}
+            gen_kubeconf_from_sa ${SERVICE_ACCOUNT} ${KUBECFG_FILE}
+        fi
+    done
+    kubectl config use-context ${PILOT_CLUSTER}
+}
+
 
 function setup_cluster() {
-  kubectl create clusterrolebinding prow-cluster-admin-binding\
-    --clusterrole=cluster-admin\
-    --user="${KUBE_USER}"
+  # use current-context if pilot_cluster not set
+  PILOT_CLUSTER=${PILOT_CLUSTER:-$(kubectl config current-context)}
+
+  unset IFS
+  k_contexts=$(kubectl config get-contexts -o name)
+  for context in ${k_contexts}; do
+     kubectl config use-context ${context}
+
+     kubectl create clusterrolebinding prow-cluster-admin-binding\
+       --clusterrole=cluster-admin\
+       --user="${KUBE_USER}"
+  done
+  if [[ "${SETUP_CLUSTERREG}" == "True" ]]; then
+      setup_clusterreg
+  fi
+  kubectl config use-context ${PILOT_CLUSTER}
+
+  if [[ "${USE_GKE}" == "True" && "${SETUP_CLUSTERREG}" == "True" ]]; then
+    ALL_CLUSTER_CIDRS=
+    for cidr in $(gcloud container clusters list --format='value(clusterIpv4Cidr)'); do
+      if [[ "${ALL_CLUSTER_CIDRS}" != "" ]]; then
+        ALL_CLUSTER_CIDRS+=','
+      fi
+      ALL_CLUSTER_CIDRS+=${cidr}
+    done
+    ALL_CLUSTER_NETTAGS=
+    for net_tag in $(gcloud compute instances list --format='value(tags.items.[0])'); do
```

Comment:
gcloud compute instances list --format='value(tags.items.[0] | sort | uniq ?

Reply:
```bash
function join_by { local IFS="$1"; shift; echo "$*"; }

ALL_CLUSTER_CIDRS=($(gcloud container clusters list --format='value(clusterIpv4Cidr)' | sort | uniq))
ALL_CLUSTER_CIDRS="${join_by , ${TEST}"
```

Reply:
Yes... `sort | uniq` is much nicer!  fixing for this and ALL_CLUSTER_NETTAGS